server/controllers/stock/lots.js

```diff
@@ -166,42 +167,130 @@ function getDupes(req, res, next) {
+    transaction.addQuery(updateStockMovement, [keepLotUuid, mergeLotUuid]);
+    transaction.addQuery(deleteLot, [mergeLotUuid]);
+  });
+
+  return transaction.execute();
+}
+
+/**
+ * GET /lots/:uuid/merge
+ *
+ * @description
+ * Merge the lots_to_merge (in the body) into the lot to keep (given by uuid).
+ * This is a accomplished in two steps for each lot to merge:
+ *   1. Replace all references to the lot to be merged with
+ *      references to the lot to keep.
+ *   2. Delete the lot to be merged
+ */
+function merge(req, res, next) {
+  const { uuid } = req.params;
+  const lotsToMerge = req.body.lotsToMerge.map(db.bid);
+
+  mergeLotsInternal(uuid, lotsToMerge)
+    .then(() => {
+      res.sendStatus(200);
+    })
+    .catch(next)
+    .done();
+}
+
+/**
+ * GET /lots/merge/auto
+ *
+ * @description
+ * Finds and merges all lots suitable for automatic merging
+ *  - To qualify, the lots must have the same inventory_uuid,
+ *    label, and expiration date.
+ */
+function autoMerge(req, res, next) {
+  // The first query gets the inventory UUID for each
+  // inventory article with duplicate lots (having the
+  // same label, inventory_uuid, and expiration_date).
+  // (Since these are grouped, only one of the several
+  // lots is given.)
+  const query1 = `
+  SELECT
+    BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+    BUID(i.uuid) AS inventory_uuid, i.text as inventory_text,
+    COUNT(*) as num_duplicates
+  FROM lot l
+  JOIN inventory i ON i.uuid = l.inventory_uuid
+  GROUP BY label, inventory_uuid, expiration_date HAVING num_duplicates > 1
+`;
+
+  // The second query gets all lots matching the label,
+  // inventory_uuid, and expiration dates
+  const query2 = `
+    SELECT
+      BUID(l.uuid) AS uuid, l.label, l.expiration_date,
+      BUID(i.uuid) AS inventory_uuid, i.text as inventory_text
+    FROM lot l
+    JOIN inventory i ON i.uuid = l.inventory_uuid
+    WHERE l.label=? AND i.uuid=? AND l.expiration_date=DATE(?)
+  `;
+
+  let numInventories = null;
+  let numLots = null;
+
+  db.exec(query1, [])
+    .then((rows) => {
+      numInventories = rows.length;
+      numLots = rows.reduce((sum, row) => {
+        return sum + row.num_duplicates;
+      }, 0) - numInventories;
+      const dbPromises = [];
+      rows.forEach((row) => {
+        const promise = db.exec(query2, [row.label, db.bid(row.inventory_uuid), row.expiration_date])
+          .then((lots) => {
+            // Arbitrarily keep the first lot and merge the duplicates into it
+            const keepLotUuid = lots[0].uuid;
+            const lotUuids = lots.reduce((list, elt) => {
+              if (elt.uuid !== keepLotUuid) {
+                list.push(elt.uuid);
+              }
+              return list;
+            }, []);
+            return mergeLotsInternal(keepLotUuid, lotUuids);
+          });
+        dbPromises.push(promise);
+      });
+      return Promise.all(dbPromises);
```

Comment:
Exactly.  All errors are now caught by the `catch(next)` statement below since they are returned to the promise chain.